test/test_dataloader.py

```diff
@@ -175,21 +226,73 @@ def __len__(self):
 
 
 def _test_timeout():
-    os.close(sys.stderr.fileno())
-    sys.stderr.close()
     dataset = SleepDataset(10, 10)
     dataloader = DataLoader(dataset, batch_size=2, num_workers=2, timeout=1)
     _ = next(iter(dataloader))
 
 
 def _test_segfault():
-    os.close(sys.stderr.fileno())
-    sys.stderr.close()
     dataset = SegfaultDataset(10)
     dataloader = DataLoader(dataset, batch_size=2, num_workers=2)
     _ = next(iter(dataloader))
 
 
+def _test_interrupt_retry(timeout=0):
+    dataset = TensorDataset(torch.randn(1, 1), torch.randn(1, 1))
+    dataloader = DataLoader(dataset, batch_size=1, num_workers=1, timeout=timeout)
+    dataloaderiter = iter(dataloader)
+
+    # make SIGUSR1 interrupt
+    def handler(signum, frame):
+        pass
+    signal.signal(signal.SIGUSR1, handler)
+
+    # Replace iterator getter with a wrapper that reliably calls an
+    # interruptable blocking recv syscall to simulate interruption during recv
+    # in queue.get.
+    # The used socket.recv call below in the replacing function is quite
+    # dangerous because it blocks everything on Python side (likely by holding
+    # GIL), including any Python signal handlers and the cleaning up in
+    # dataloder.__del__ when this process exits. To prevent orphan worker child,
+    # we manually terminate worker process here.
+    # Conveniently, the worker has SIGTERM handler installed so SIGTERM from
+    # loader process won't cause loader error.
+
+    data_queue = dataloaderiter.data_queue
+    data = data_queue.get()  # ensure that worker handlers are installed
+    for w in dataloaderiter.workers:
+        w.terminate()
+
+    def interruptable_get(*args, **kwargs):
+        if dataloaderiter.shutdown:
+            return data
+        # get and config timeout if the argument is present
+        if timeout >= 0:
+            if 'timeout' in kwargs:
+                timeout_val = kwargs['timeout']
+            elif len(args) > 1:
+                timeout_val = args[1]  # first arg is `block`
+            else:
+                timeout_val = None
+            socket.setdefaulttimeout(timeout_val)
+        s = socket.socket(socket.AF_INET, type=socket.SOCK_DGRAM)
+        s.bind(("127.0.0.1", 0))
+        try:
+            return s.recv(1024)
+        except socket.timeout:
+            raise queue.Empty
+        finally:
+            s.close()
+
+    try:
+        if isinstance(data_queue, QueueWrapper):
+            data_queue.queue.get = interruptable_get
+    except Exception:
+        # in Python >= 3.5 or on Windows QueueWrapper is not a class but a no-op
+        pass
```

Comment:
It's not really an exception, can you just use an `if` with `sys.version_info`? Also, why do we need to replace `get` with our custom implementation? I thought it is interruptible in its default implementation and we're actually trying to see if we can improve this

Reply:
The default get is interruptible, but only during its `recv` call. And it doesn't call `recv` unless the queue has data. So in reality there is only a tiny timeframe where it is interruptible, so it is really not testable. The custom implementation can be interrupted at any time.

Reply:
Makes sense. Can you still change the code to use an `if` instead of raising exceptions?

Reply:
Also, why can't we patch `get` in Python 3.5? Should we just skip this test in this version, since it doesn't suffer from the same problems

Reply:
`get` is already patched by Python in 3.5. But yeah, I'll add an `if` here.